File: mcp_server_development/python/src/session10/verify_good.py
"""Python 版の動作確認と、ツール定義サイズの計測

実行： docker compose exec python python src/session10/verify_good.py

クライアント側のスクリプトなので print() を使ってかまいません。
"""

from __future__ import annotations

import json
from typing import Any

import anyio
from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


def estimate_tokens(text: str) -> int:
    """TypeScript 版（tokens.ts）と同じ近似モデル"""
    ascii_count = sum(1 for ch in text if ord(ch) < 128)
    wide_count = len(text) - ascii_count
    return -(-ascii_count // 4) + wide_count


def payload(result: Any) -> dict[str, Any]:
    """structured_content を持つ結果から辞書を取り出す

    mcp 2.0 のモデルは snake_case です（TypeScript SDK の structuredContent に相当）。
    """
    if result.structured_content is not None:
        return result.structured_content
    return json.loads(result.content[0].text)


async def main() -> None:
    params = StdioServerParameters(command="python", args=["src/session10/good_server.py"])

    async with stdio_client(params) as (read, write):
        async with ClientSession(read, write) as session:
            init = await session.initialize()
            tools = (await session.list_tools()).tools
            names = sorted(tool.name for tool in tools)
            print(
                f"[1/5] 接続: {init.server_info.name} v{init.server_info.version} "
                f"/ tools={', '.join(names)}"
            )

            found = payload(await session.call_tool("search_requests", {"status": ["in_review"]}))
            ids = ", ".join(item["id"] for item in found["items"])
            print(f"[2/5] search_requests(status=['in_review']): total={found['total']} ids={ids}")

            detail = (await session.call_tool("get_request", {"request_id": "req-1003"})).content[
                0
            ].text
            print(
                f"[3/5] get_request(req-1003): 承認ルート="
                f"{detail.count('（u-9')} 段 / コメント="
                f"{'含まれる' if 'コメント（' in detail else '含まれない'}"
            )

            dry = payload(
                await session.call_tool(
                    "decide_request", {"request_id": "req-1003", "decision": "approve"}
                )
            )
            print(
                f"[4/5] decide_request(ドライラン): applied={dry['applied']} "
                f"{dry['stepLabel']} 確定後={dry['nextStatus']} "
                f"審査終了={'はい' if dry['finalizes'] else 'いいえ'}"
            )

            applied = payload(
                await session.call_tool(
                    "decide_request",
                    {
                        "request_id": "req-1003",
                        "decision": "approve",
                        "confirm": True,
                        "preview_token": dry["previewToken"],
                    },
                )
            )
            print(
                f"[5/5] decide_request(確定): applied={applied['applied']} "
                f"{applied['stepLabel']} 状態={applied['status']}"
            )

            total_chars = 0
            total_tokens = 0
            for tool in sorted(tools, key=lambda t: t.name):
                text = json.dumps(
                    tool.model_dump(mode="json", exclude_none=True),
                    ensure_ascii=False,
                    separators=(",", ":"),
                )
                total_chars += len(text)
                total_tokens += estimate_tokens(text)
                print(f"  {tool.name:<20}{len(text):>8}{estimate_tokens(text):>9}")
            print(f"[定義サイズ] {len(tools)} 本 / {total_chars} chars / 約 {total_tokens} tokens")
            print("OK: Python 版の 3 ツールが動作しました")


if __name__ == "__main__":
    anyio.run(main)

File: mcp_server_development/node/src/session10/verify-good.ts
/**
 * Good 実装の動作確認
 *
 * 実行： docker compose exec node npx tsx src/session10/verify-good.ts
 *
 * クライアント側のスクリプトなので console.log を使ってかまいません
 * （禁止されているのは「サーバープロセスの stdout」だけです）。
 */
import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { InMemoryTransport } from "@modelcontextprotocol/sdk/inMemory.js";

import { createWorkflowServer } from "./good-create-server.js";

/** ツール結果の 1 つめのテキストを取り出す */
function text(result: { content?: unknown; structuredContent?: unknown; isError?: boolean; [key: string]: unknown }): string {
  const first = (result.content as Array<{ type: string; text?: string }> | undefined)?.[0];
  return first?.text ?? "";
}

/** ドライランの出力から previewToken を拾う */
function pickToken(message: string): string {
  return /previewToken: (\S+)/.exec(message)?.[1] ?? "";
}

const { server } = createWorkflowServer();
const [clientTransport, serverTransport] = InMemoryTransport.createLinkedPair();
const client = new Client({ name: "session10-verify", version: "1.0.0" });
await server.connect(serverTransport);
await client.connect(clientTransport);

const names = (await client.listTools()).tools.map((tool) => tool.name).sort();
console.log(`[1/9] tools/list: ${names.length} 本 = ${names.join(", ")}`);

const searched = await client.callTool({
  name: "search_requests",
  arguments: { status: ["in_review"] },
});
const structured = searched.structuredContent as {
  total: number;
  items: Array<{ id: string }>;
};
console.log(
  `[2/9] search_requests(status=["in_review"]): total=${structured.total} ids=${structured.items
    .map((item) => item.id)
    .join(", ")}`,
);

const detail = text(
  await client.callTool({
    name: "get_request",
    arguments: { requestId: "req-1003", include: ["comments"] },
  }),
);
console.log(
  `[3/9] get_request(req-1003, include=["comments"]): ` +
    `コメント=${/コメント（(\d+) 件）/.exec(detail)?.[1] ?? "0"} 件 / ` +
    `添付=${detail.includes("添付（") ? "含まれる" : "含まれない"} / ` +
    `履歴=${detail.includes("履歴（") ? "含まれる" : "含まれない"}`,
);

const created = text(
  await client.callTool({
    name: "save_request",
    arguments: {
      title: "書籍購入（技術書 3 冊）",
      category: "purchase",
      amountYen: 12_000,
      body: "チームの技術書を 3 冊購入します。",
    },
  }),
);
const newId = /^(req-\d{4})/.exec(created)?.[1] ?? "";
console.log(`[4/9] save_request(新規): ${created.split("\n")[0]}`);

const updated = text(
  await client.callTool({
    name: "save_request",
    arguments: { requestId: newId, amountYen: 24_000 },
  }),
);
console.log(`[5/9] save_request(更新): ${updated.split("\n")[0]}`);

const submitDry = text(
  await client.callTool({ name: "submit_request", arguments: { requestId: newId } }),
);
console.log(
  `[6/9] submit_request(ドライラン): ${submitDry.split("\n")[1]} / previewToken=${
    pickToken(submitDry) === "" ? "なし" : "あり"
  }`,
);

const submitted = text(
  await client.callTool({
    name: "submit_request",
    arguments: { requestId: newId, confirm: true, previewToken: pickToken(submitDry) },
  }),
);
console.log(`[7/9] submit_request(確定): ${submitted}`);

const decideDry = await client.callTool({
  name: "decide_request",
  arguments: { requestId: "req-1003", decision: "approve" },
});
const dryStructured = decideDry.structuredContent as {
  applied: boolean;
  stepLabel: string;
  nextStatus: string;
  finalizes: boolean;
  previewToken?: string;
};
console.log(
  `[8/9] decide_request(ドライラン): applied=${dryStructured.applied} ` +
    `${dryStructured.stepLabel} 確定後=${dryStructured.nextStatus} ` +
    `審査終了=${dryStructured.finalizes ? "はい" : "いいえ"}`,
);

const decided = await client.callTool({
  name: "decide_request",
  arguments: {
    requestId: "req-1003",
    decision: "approve",
    confirm: true,
    previewToken: dryStructured.previewToken,
  },
});
console.log(`[9/9] decide_request(確定): ${text(decided).split("\n")[0]}`);

// ── 異常系（二段階が本当に効いているかの確認） ──
const noToken = await client.callTool({
  name: "decide_request",
  arguments: { requestId: "req-1002", decision: "approve", confirm: true },
});
console.log(`[異常系1] previewToken なしで confirm: isError=${noToken.isError === true}`);

const staleDry = text(
  await client.callTool({
    name: "decide_request",
    arguments: { requestId: "req-1002", decision: "approve" },
  }),
);
// プレビュー後に申請が変わったことを再現する（コメントを 1 件付ける）
await client.callTool({
  name: "comment_on_request",
  arguments: { requestId: "req-1002", body: "追加の見積を確認しました。" },
});
const stale = await client.callTool({
  name: "decide_request",
  arguments: {
    requestId: "req-1002",
    decision: "approve",
    confirm: true,
    previewToken: pickToken(staleDry),
  },
});
console.log(`[異常系2] 申請が変わった後の previewToken: isError=${stale.isError === true}`);

await client.close();
await server.close();
console.log("OK: セッション10 の Good 実装は仕様どおりに動作しています");
